cruise-control-metrics-reporter/src/main/java/com/linkedin/kafka/cruisecontrol/metricsreporter/CruiseControlMetricsReporter.java

```diff
@@ -224,6 +235,30 @@ protected NewTopic createMetricsTopicFromReporterConfig(CruiseControlMetricsRepo
     return newTopic;
   }
 
+  protected void createCruiseControlMetricsProducer(Properties producerProps) throws KafkaException {
+    CruiseControlMetricsUtils.retry(() -> {
+      try {
+        _producer = new KafkaProducer<>(producerProps);
+        return false;
+      } catch (KafkaException e) {
+        if (e.getCause() instanceof ConfigException) {
+          // Check if the config exception is caused by bootstrap.servers config
+          try {
+            ProducerConfig config = new ProducerConfig(producerProps);
+            ClientUtils.parseAndValidateAddresses(
```

Comment:
I'd run inside the `retry` only  the address validation and execute the `KafkaProducer` creation outside of the `retry` as we can assume that if the address validation succeeded after a couple of retries the bootstrap.servers config is valid and DNS resolution works for it.

If the creation of `KafkaProducer` throws an error when the bootstrap servers config is valid than that's caused by some other config error which probably is not a transient error that would go away after a couple of retries.

Reply:
I'm not sure I agree with this approach because I don't know how common it is to have this DNS resolution race condition. I don't want to add error-checking code in the happy path if this DNS issue only happens 5% of the time. 

I think its better to try to create the kafka producer first and then if there is a validation error for the bootstrap servers we can retry a couple of times.

Reply:
What was in my mind is that the best practice is to validate input data and config before any processing. e.g in this case validate addresses and create `KafkaProducer` only if validation passes.